src/ucs/datastruct/ptr_array.h

```diff
@@ -120,13 +149,173 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
                     (UCS_V_INITIALIZED(_var), 0) : \
                     !__ucs_ptr_array_is_free(_var = (void*)((_ptr_array)->start[_index])))
 
+/**
+ * Check if element is free, and calculate how many free elements are ahead.
+ *
+ * @param [in]  _elem      An element in the ptr array.
+ * @param [out] _size_elem Pointer to the returned number of free elements ahead.
+ *
+ * @return 1 if the element is free and 0 if it's occupied.
+ */
+static inline int
+__ucs_ptr_array_is_free_and_get_step_size(ucs_ptr_array_elem_t _elem,
+                                          uint32_t *_size_elem)
+{
+    int _is_free = ((uintptr_t)(_elem) & UCS_PTR_ARRAY_FLAG_FREE);
+
+    if (_is_free) {
+       *_size_elem = _elem >> UCS_PTR_ARRAY_SIZE_FREE_SHIFT;
+    } else {
+       *_size_elem = 1;
+    }
+    ucs_assert(*_size_elem > 0);
+
+    return _is_free;
+}
 
 /**
  * Iterate over all valid elements in the array.
+ *
+ * @param [out] _var        Pointer to current array element in the foreach.
+ * @param [out] _index      Index variable to use as iterator (unsigned).
+ * @param [in]  _ptr_array  Pointer to a ptr array.
  */
 #define ucs_ptr_array_for_each(_var, _index, _ptr_array) \
-    for (_index = 0; _index < (_ptr_array)->size; ++_index) \
-         if (!__ucs_ptr_array_is_free(_var = (void*)((_ptr_array)->start[_index]))) \
+    uint32_t step_size = 1; \
```

Comment:
we don't need `step_size` and it's bad to declare variable in this macro, since it can break some strict compilers that would report violation C standard "mixing code and declarations"

`__ucs_ptr_array_is_free_and_get_step_size()` is used only here please, make `__ucs_ptr_array_is_free_and_get_step_size()` -> `__ucs_ptr_array_is_free()`

Reply:
i.e. can just use `__ucs_ptr_array_is_free()` macro that already in ptr_array.h

Reply:
Hi @dmitrygx , 
The thing is that we need the step size in order to optimize the foreach loop - Let me think how to resolve :)

Reply:
Ah, I see...
yes, please, try to fix it

Reply:
I'd propose something:
```
static inline int
__ucs_ptr_array_is_free_and_update_iter(ucs_ptr_array_elem_t _elem,
                                          unsigned *iter)
{
    int _is_free = ((uintptr_t)_elem) & UCS_PTR_ARRAY_FLAG_FREE);

    if (_is_free) {
       *iter += _elem >> UCS_PTR_ARRAY_SIZE_FREE_SHIFT;
    } else {
       *iter += 1;
    }
    ucs_assert(*_size_elem > 0);

    return is_free;
}
...

#define ucs_ptr_array_for_each(_var, _index, _ptr_array) \
    for ((_index) = 0; (_index) < (_ptr_array)->size; ) \
        if ((ucs_likely(!__ucs_ptr_array_is_free_and_update_iter(
                             (ucs_ptr_array_elem_t)((_var) = (void *)((_ptr_array)->start[_index])), \
                             &(_index)))))
```
also, no need for the `\` at the last line

Reply:
@dmitrygx , Regarding the '\', now that I'm thinking about it again it's unnecessary, I agree.

Also, I implemented your recommendation for the iterator update but it doesn't work, probably since the _index type is set by the user (sometimes it's an int and sometimes it's an unsigned - it will be very hard to write a function that handles all these types if it's a pointer) and perhaps sometimes the user may set it to a type with a size smaller than 32-bit which will then cause issues.

I recommend to keep it with the variable - but I did rename the variable as follows:
{code}
#define ucs_ptr_array_for_each(_var, _index, _ptr_array) \
    uint32_t __foreach_step_size;\
    for ((_index) = 0; ((_index) < (_ptr_array)->size); (_index) += __foreach_step_size) \
       if ((ucs_likely(!__ucs_ptr_array_is_free_and_get_step_size( \
                            (ucs_ptr_array_elem_t)((_var) = (void *)((_ptr_array)->start[(_index)])), \
                            &(__foreach_step_size)))))

{code}

Reply:
> I recommend to keep it with the variable - but I did rename the variable as follows:

unfortunately, it will break some compilers. Not sure we can allow it.

> Also, I implemented your recommendation for the iterator update but it doesn't work

why it doesn't work? or you mean it would not work potentially due to described limitations?
Maybe we should require from user always pass it as an `unsigned` (i.e. the same type as we have for `index`)

Reply:
OK, @dmitrygx  - Please review the latest code.
Following a discussion with @alex--m , please see a nice and elegant solution for most concerns and comment.